File: apps/web/lib/api/laboratory.ts
/**
 * Laboratory batches, over HTTP.
 *
 * Parsed rather than cast, for the reason set out in `materials.ts`: a
 * server that renamed a field would hand back rows whose value is
 * `undefined`, and the grid would render blanks that look exactly like a
 * database with nothing recorded.
 *
 * 🔴 THE MASSES ARE STRINGS, AND THAT IS A CORRECTNESS REQUIREMENT.
 *
 * `planned_quantity_kg` is `NUMERIC(14,4)` and `tolerance_percent` is
 * `NUMERIC(6,4)`. FastAPI's `jsonable_encoder` maps `Decimal` to
 * **float** — measured, `Decimal("12.5000") -> 12.5` — so until this was
 * wired the API was shipping a controlled batch mass with its stored
 * scale destroyed. `CLAUDE.md` §5: *"NUMERIC, never float, for
 * percentages, masses, densities and measured values."*
 *
 * That was fixed server-side (`_decimal_strings` in
 * `app/domains/laboratory/service.py`, and
 * `tests/test_laboratory_testing_serialisation.py` pins it). Typing them
 * as `z.string()` here is the client half of the same contract: if the
 * server ever regresses to a float, these rows fail to parse and the
 * screen says so, rather than silently displaying a rounded mass.
 *
 * **Do not `Number()` these.** No arithmetic happens in the browser —
 * §4 keeps derivation on the server, and `0.1 + 0.2` is why.
 *
 * 🔴 WHAT THE LIST ENDPOINT DOES **NOT** RETURN
 *
 * No components, no weighings, no process parameters, no deviations
 * beyond a COUNT. `list_batches` returns the batch's own columns plus
 * four sub-counts, deliberately — a queue of forty batches must not run
 * forty sub-queries. The weigh-up sheet belongs to the batch detail
 * screen, which `fetchBatch` below now serves.
 *
 * ⚠️ AND THE DETAIL RESPONSE IS NOT A SUPERSET OF THE LIST ROW. Measured
 * against the running service rather than inferred: `get_batch` returns the
 * batch's own columns plus `components`, `process_parameters`, `deviations`
 * and `samples`, and it does **not** carry `formula_code`, `formula_name` or
 * `version_code` — those come from joins only `list_batches` performs. One
 * shared schema would therefore be wrong in one direction or the other,
 * which is why there are two.
 */

import { z } from "zod";

import { apiRequest, type ApiCredentials } from "./client";

export const batchSchema = z.object({
  id: z.string(),
  batch_number: z.string(),
  status: z.string(),
  // NOT NULL in the schema, so required here. Strings — see the header.
  planned_quantity_kg: z.string(),
  tolerance_percent: z.string(),
  project_id: z.string(),
  formula_version_id: z.string(),
  // From the joins to `formula_versions` and `formulas`. Both are inner
  // joins on NOT NULL foreign keys, so these are always present.
  version_code: z.string(),
  formula_code: z.string(),
  formula_name: z.string(),
  // A batch that has not started has no `started_at`. Nullable, but the
  // key is always sent, so `.nullable()` and never `.optional()` — the
  // distinction Codex flagged on `target_release_date`.
  started_at: z.string().nullable(),
  completed_at: z.string().nullable(),
  updated_at: z.string(),
  // Sub-counts. Integers, so numbers rather than strings: they are
  // cardinalities, not measurements, and no scale can be lost.
  component_count: z.number(),
  unweighed_count: z.number(),
  deviation_count: z.number(),
  sample_count: z.number(),
});

export type Batch = z.infer<typeof batchSchema>;

const batchList = z.array(batchSchema);

export function fetchBatches(
  credentials: ApiCredentials,
  signal?: AbortSignal,
): Promise<Batch[]> {
  return apiRequest(
    { path: "/api/laboratory/batches", credentials, signal },
    (payload) => batchList.parse(payload),
  );
}

// ---------------------------------------------------------------------------
// The batch detail — the weigh-up sheet and everything recorded against it
// ---------------------------------------------------------------------------

/**
 * One line of the weigh-up sheet.
 *
 * 🔴 `deviation` IS `null` WHEN THE LINE IS UNWEIGHED, AND THAT IS NOT ZERO.
 * The server is explicit about it — *"NOT a zero deviation. An unweighed line
 * is unweighed, and reporting it as 0.00% within tolerance would make an
 * incomplete batch look finished."* The screen must render the two
 * differently or it re-creates that defect one layer up.
 *
 * Masses are strings, per this file's header. Do not `Number()` them: the
 * server computes the deviation and sends it already decided.
 */
export const batchComponentSchema = z.object({
  id: z.string(),
  material_id: z.string(),
  material_code: z.string(),
  material_name: z.string(),
  role: z.string().nullable(),
  display_order: z.number(),
  planned_mass_kg: z.string(),
  actual_mass_kg: z.string().nullable(),
  weighed_at: z.string().nullable(),
  notes: z.string().nullable(),
  material_lot_id: z.string().nullable(),
  lot_number: z.string().nullable(),
  lot_status: z.string().nullable(),
  deviation: z
    .object({
      delta_kg: z.string(),
      delta_percent: z.string(),
      within_tolerance: z.boolean(),
    })
    .nullable(),
});

export type BatchComponent = z.infer<typeof batchComponentSchema>;

export const processParameterSchema = z.object({
  id: z.string(),
  parameter_code: z.string(),
  value: z.string(),
  unit: z.string(),
  stage: z.string().nullable(),
  recorded_at: z.string(),
  notes: z.string().nullable(),
});

export type ProcessParameter = z.infer<typeof processParameterSchema>;

export const batchDeviationSchema = z.object({
  id: z.string(),
  description: z.string(),
  severity: z.string(),
  raised_at: z.string(),
  resolution: z.string().nullable(),
  resolved_at: z.string().nullable(),
  batch_component_id: z.string().nullable(),
});

export type BatchDeviation = z.infer<typeof batchDeviationSchema>;

export const batchSampleSchema = z.object({
  id: z.string(),
  sample_number: z.string(),
  quantity_g: z.string().nullable(),
  purpose: z.string().nullable(),
  status: z.string(),
  storage_location: z.string().nullable(),
  taken_at: z.string(),
  expires_on: z.string().nullable(),
});

export type BatchSample = z.infer<typeof batchSampleSchema>;

export const batchDetailSchema = z.object({
  id: z.string(),
  organization_id: z.string(),
  project_id: z.string(),
  formula_version_id: z.string(),
  batch_number: z.string(),
  status: z.string(),
  planned_quantity_kg: z.string(),
  tolerance_percent: z.string(),
  purpose: z.string().nullable(),
  mixing_procedure: z.string().nullable(),
  notes: z.string().nullable(),
  created_by: z.string(),
  created_at: z.string(),
  authorized_by: z.string().nullable(),
  authorized_at: z.string().nullable(),
  executed_by: z.string().nullable(),
  started_at: z.string().nullable(),
  completed_at: z.string().nullable(),
  reviewed_by: z.string().nullable(),
  reviewed_at: z.string().nullable(),
  review_note: z.string().nullable(),
  updated_at: z.string(),
  components: z.array(batchComponentSchema),
  process_parameters: z.array(processParameterSchema),
  deviations: z.array(batchDeviationSchema),
  samples: z.array(batchSampleSchema),
});

export type BatchDetail = z.infer<typeof batchDetailSchema>;

export function fetchBatch(
  credentials: ApiCredentials,
  batchId: string,
  signal?: AbortSignal,
): Promise<BatchDetail> {
  return apiRequest(
    { path: `/api/laboratory/batches/${batchId}`, credentials, signal },
    (payload) => batchDetailSchema.parse(payload),
  );
}

// ---------------------------------------------------------------------------
// The lifecycle — each of these is a step somebody performs at the bench
// ---------------------------------------------------------------------------
//
// 🔴 NONE OF THESE PARSE THE RESPONSE INTO A BATCH, AND THAT IS DELIBERATE.
// The routes return varying shapes (a batch row, a deviation result, a bare
// `{id}`), and the screen refetches the batch after every one of them rather
// than patching local state from a reply. §10's `status`, `display_color` and
// `final_status` are DERIVED and server-owned; a client that set any of them
// optimistically would be inventing a safety-critical field. Refetch is
// slower and it is the only version that cannot be wrong.

/** Issue the weigh-up sheet. Planned quantities freeze here. */
export function authorizeBatch(
  credentials: ApiCredentials,
  batchId: string,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/authorization`,
      method: "POST",
      body: {},
      credentials,
    },
    (payload) => payload,
  );
}

/** Begin execution. */
export function startBatch(
  credentials: ApiCredentials,
  batchId: string,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/start`,
      method: "POST",
      body: {},
      credentials,
    },
    (payload) => payload,
  );
}

export interface WeighingRequest {
  readonly actual_mass_kg: string;
  /** Omitted, never sent as "", when the bench recorded no lot. */
  readonly material_lot_id?: string;
}

/**
 * Record what was actually weighed.
 *
 * The route returns the deviation immediately, and its docstring says why:
 * *"a technician needs to know at the bench, while the material is still in
 * front of them, not at review a day later."* The screen surfaces it and then
 * refetches, so the sheet and the answer come from the same read.
 */
export function recordWeighing(
  credentials: ApiCredentials,
  batchId: string,
  componentId: string,
  request: WeighingRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/components/${componentId}/weighing`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

export interface ProcessParameterRequest {
  readonly parameter_code: string;
  readonly value: string;
  readonly unit: string;
  readonly stage?: string;
  readonly notes?: string;
}

/** Mixing RPM, mixing time, temperature, vacuum — value + unit, never a string. */
export function recordProcessParameter(
  credentials: ApiCredentials,
  batchId: string,
  request: ProcessParameterRequest,
): Promise<{ id: string }> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/process-parameters`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => z.object({ id: z.string() }).parse(payload),
  );
}

export interface DeviationRequest {
  readonly description: string;
  readonly severity: string;
  readonly batch_component_id?: string;
}

export function raiseDeviation(
  credentials: ApiCredentials,
  batchId: string,
  request: DeviationRequest,
): Promise<{ id: string }> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/deviations`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => z.object({ id: z.string() }).parse(payload),
  );
}

export interface SampleRequest {
  readonly sample_number: string;
  readonly quantity_g?: string;
  readonly purpose?: string;
  readonly storage_location?: string;
  readonly notes?: string;
}

/** Take a sample. This is the record every future test result cites. */
export function createSample(
  credentials: ApiCredentials,
  batchId: string,
  request: SampleRequest,
): Promise<{ id: string }> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/samples`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => z.object({ id: z.string() }).parse(payload),
  );
}

/** Close execution. The server refuses while any line is unweighed. */
export function completeBatch(
  credentials: ApiCredentials,
  batchId: string,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/completion`,
      method: "POST",
      body: {},
      credentials,
    },
    (payload) => payload,
  );
}

/**
 * Chemist Review — accept for testing, or reject for process deviation.
 *
 * ⚠️ THE TWO DECISIONS REQUIRE DIFFERENT PERMISSIONS SERVER-SIDE. Only the
 * Engineer holds `batch.reject` (`REVIEW_PERMISSION` in the route). The screen
 * offers both and lets the server refuse, per §6: a frontend permission check
 * is cosmetic and the server is authoritative. Hiding the button here would
 * also need `/api/me` to report permissions, which it does not — I79.
 */
export interface ReviewRequest {
  readonly decision: "accept" | "reject";
  readonly note?: string;
}

export function reviewBatch(
  credentials: ApiCredentials,
  batchId: string,
  request: ReviewRequest,
): Promise<unknown> {
  return apiRequest(
    {
      path: `/api/laboratory/batches/${batchId}/review`,
      method: "POST",
      body: request,
      credentials,
    },
    (payload) => payload,
  );
}

File: apps/api/app/api/laboratory.py
"""Laboratory routes — the guided batch flow.

**Each step is guarded by the permission that names it**, and those
permissions were checked against migration 002 before this module was
written rather than invented to fit it:

    batch.create    Chemist            create the batch, issue the sheet
    batch.execute   Technician         start it, weigh, capture process data
    batch.complete  Engineer, Tech     close execution
    batch.reject    Engineer           reject at review
    sample.create   Chemist, Eng, Tech take a sample

There is deliberately no `batch.authorize`: no such permission exists, and
inventing one would have produced a control no role holds — the defect
migration 016 had to close for `material.approve_production`. The
authorising act is the Lead's `formula.approve_lab`, without which the
batch cannot be created at all.

**Review is one endpoint, two permissions.** Accepting needs
`batch.complete`; rejecting needs `batch.reject`, which only the Engineer
holds. Guarding the endpoint with either alone would have handed the
Technician a rejection right the permission model withholds, or blocked
the Engineer from accepting. Resolved from the decision in the body, the
same shape as the material status route.
"""

from __future__ import annotations

import uuid
from decimal import Decimal
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

# 🔴 THE READS GO THROUGH THE ORCHESTRATOR (§0.2, I103).
#
# This module called its domain service directly, and the department
# conductor written for it had NO CALLERS -- a layer nothing reaches is the
# same defect as a route nothing calls, which is a rule this repository
# already has. `app/api/dashboards.py` was wired first; this is the same
# move.
#
# ⚠️ THE WRITES DELIBERATELY DO NOT. §4: humans approve, and AI must not
# authorize a batch, confirm a test or move a result from YELLOW to GREEN.
# So the orchestrator exposes no write-side entry point at all, and every
# mutation below still calls the domain service directly. The asymmetry is
# the rule, not an omission -- if a write ever appears on that door, it is a
# §4 violation and not a convenience.
#
# ⚠️ `require_permission(...)` ON EACH ROUTE STAYS. The conductor asserts the
# same permission; that is defence in depth. The dependency refuses an
# unauthenticated caller before any handler runs, and the conductor refuses on
# the paths that have no route.
from app.agents.orchestrators.root_orchestrator import (
    laboratory_batch,
    laboratory_batches,
)
from app.core.security import (
    PermissionDenied,
    Principal,
    get_db,
    get_principal,
    require_permission,
)
from app.core.tenancy import CrossTenantReferenceError
from app.domains.laboratory.service import (
    BatchError,
    BatchInput,
    BatchNotFoundError,
    BatchStateError,
    DeviationInput,
    LaboratoryError,
    ProcessParameterInput,
    SampleInput,
    authorize_batch,
    complete_batch,
    create_batch,
    create_sample,
    raise_deviation,
    record_process_parameter,
    record_weighing,
    review_batch,
    start_batch,
)

router = APIRouter()

__all__ = ["router"]

# Which permission each review decision requires. A table for the same
# reason the material lifecycle has one: the mapping IS the authorization
# model for this step and must be readable without following control flow.
REVIEW_PERMISSION: dict[str, str] = {
    "accept": "batch.complete",
    "reject": "batch.reject",
}


class BatchCreate(BaseModel):
    formula_version_id: uuid.UUID
    batch_number: str = Field(min_length=3, max_length=50)
    # `Decimal`, never `float`. A batch quantity is a controlled mass and
    # the engine refuses a float at its boundary; declaring it as a float
    # here would undo that at the one point a number enters the system.
    planned_quantity_kg: Decimal = Field(gt=0)
    tolerance_percent: Decimal | None = Field(default=None, ge=0, le=100)
    purpose: str | None = Field(default=None, max_length=500)
    mixing_procedure: str | None = None
    notes: str | None = None


class WeighingCreate(BaseModel):
    actual_mass_kg: Decimal = Field(ge=0)
    material_lot_id: uuid.UUID | None = None


class ProcessParameterCreate(BaseModel):
    parameter_code: str = Field(min_length=1, max_length=50)
    value: Decimal
    # Required, not optional. A bare number labelled `temperature` cannot
    # be compared against another batch, and CLAUDE.md §5 requires
    # measurements stored as value + unit.
    unit: str = Field(min_length=1, max_length=20)
    stage: str | None = Field(default=None, max_length=50)
    notes: str | None = None


class DeviationCreate(BaseModel):
    description: str = Field(min_length=3, max_length=2000)
    severity: str = Field(default="minor", pattern="^(minor|major|critical)$")
    batch_component_id: uuid.UUID | None = None


class SampleCreate(BaseModel):
    sample_number: str = Field(min_length=3, max_length=50)
    quantity_g: Decimal | None = Field(default=None, gt=0)
    purpose: str | None = Field(default=None, max_length=500)
    storage_location: str | None = Field(default=None, max_length=200)
    notes: str | None = None


class ReviewCreate(BaseModel):
    decision: str = Field(pattern="^(accept|reject)$")
    note: str | None = Field(default=None, max_length=2000)


def _missing(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


def _conflict(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc))


def _invalid(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc))


@router.get("", tags=["laboratory"])
def get_batches(
    project_id: uuid.UUID | None = Query(default=None),
    status_filter: str | None = Query(default=None, alias="status"),
    principal: Principal = Depends(require_permission("batch.view")),
    session: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """The batch queue.

    `unweighed_count` comes back on every row because it is what makes the
    queue actionable — CLAUDE.md §11 requires counts to represent items
    needing action, not total rows.
    """
    return laboratory_batches(
        session,
        organization_id=principal.organization_id,
        permissions=principal.permissions,
        project_id=project_id,
        status=status_filter,
    )


@router.post("", status_code=status.HTTP_201_CREATED, tags=["laboratory"])
def post_batch(
    payload: BatchCreate,
    principal: Principal = Depends(require_permission("batch.create")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Create a batch and calculate its weigh-up sheet.

    A 409 means the formula version exists but has not been approved for
    laboratory trial. That distinction matters to the caller: the first is
    a mistake, the second is a step somebody still has to take.
    """
    try:
        return create_batch(
            session,
            formula_version_id=payload.formula_version_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=BatchInput(
                batch_number=payload.batch_number,
                planned_quantity_kg=payload.planned_quantity_kg,
                tolerance_percent=payload.tolerance_percent,
                purpose=payload.purpose,
                mixing_procedure=payload.mixing_procedure,
                notes=payload.notes,
            ),
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc
    except (BatchError, CrossTenantReferenceError) as exc:
        raise _invalid(exc) from exc


@router.get("/{batch_id}", tags=["laboratory"])
def get_one_batch(
    batch_id: uuid.UUID,
    principal: Principal = Depends(require_permission("batch.view")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """The sheet, what was weighed against it, process data, deviations,
    samples — and each line's deviation computed at read time."""
    try:
        return laboratory_batch(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            permissions=principal.permissions,
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc


@router.post("/{batch_id}/authorization", tags=["laboratory"])
def post_authorization(
    batch_id: uuid.UUID,
    principal: Principal = Depends(require_permission("batch.create")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Issue the weigh-up sheet. Planned quantities freeze here."""
    try:
        return authorize_batch(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc


@router.post("/{batch_id}/start", tags=["laboratory"])
def post_start(
    batch_id: uuid.UUID,
    principal: Principal = Depends(require_permission("batch.execute")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    try:
        return start_batch(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc


@router.post("/{batch_id}/components/{component_id}/weighing", tags=["laboratory"])
def post_weighing(
    batch_id: uuid.UUID,
    component_id: uuid.UUID,
    payload: WeighingCreate,
    principal: Principal = Depends(require_permission("batch.execute")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Record what was actually weighed, and from which lot.

    Returns the deviation immediately — delta, percentage and whether it
    is within the batch's tolerance — because a technician needs to know
    at the bench, while the material is still in front of them, not at
    review a day later.
    """
    try:
        return record_weighing(
            session,
            batch_id=batch_id,
            component_id=component_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            actual_mass_kg=payload.actual_mass_kg,
            material_lot_id=payload.material_lot_id,
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc
    except LaboratoryError as exc:
        raise _invalid(exc) from exc


@router.post(
    "/{batch_id}/process-parameters", status_code=status.HTTP_201_CREATED, tags=["laboratory"]
)
def post_process_parameter(
    batch_id: uuid.UUID,
    payload: ProcessParameterCreate,
    principal: Principal = Depends(require_permission("batch.execute")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    """Capture mixing RPM, mixing time, temperature, vacuum — value + unit."""
    try:
        parameter_id = record_process_parameter(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=ProcessParameterInput(**payload.model_dump()),
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc
    return {"id": str(parameter_id)}


@router.post("/{batch_id}/deviations", status_code=status.HTTP_201_CREATED, tags=["laboratory"])
def post_deviation(
    batch_id: uuid.UUID,
    payload: DeviationCreate,
    # Either the person at the bench or the one reviewing may raise one.
    # A deviation noticed at review is the evidence the review exists to
    # act on, and refusing it there would push it onto paper.
    principal: Principal = Depends(require_permission("batch.execute", "batch.complete")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    try:
        deviation_id = raise_deviation(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=DeviationInput(**payload.model_dump()),
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    return {"id": str(deviation_id)}


@router.post("/{batch_id}/samples", status_code=status.HTTP_201_CREATED, tags=["laboratory"])
def post_sample(
    batch_id: uuid.UUID,
    payload: SampleCreate,
    principal: Principal = Depends(require_permission("sample.create")),
    session: Session = Depends(get_db),
) -> dict[str, str]:
    """Take a sample. This is the record every future test result cites."""
    try:
        sample_id = create_sample(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            spec=SampleInput(**payload.model_dump()),
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc
    except BatchError as exc:
        raise _invalid(exc) from exc
    return {"id": str(sample_id)}


@router.post("/{batch_id}/completion", tags=["laboratory"])
def post_completion(
    batch_id: uuid.UUID,
    principal: Principal = Depends(require_permission("batch.complete")),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Close execution. Refused while any line is unweighed."""
    try:
        return complete_batch(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc


@router.post("/{batch_id}/review", tags=["laboratory"])
def post_review(
    batch_id: uuid.UUID,
    payload: ReviewCreate,
    principal: Principal = Depends(get_principal),
    session: Session = Depends(get_db),
) -> dict[str, Any]:
    """Chemist Review: Accept for Testing, or Reject for Process Deviation.

    Depends on `get_principal` because the required permission depends on
    the DECISION: only the Engineer holds `batch.reject`. A single
    permission on the endpoint would either hand the Technician a
    rejection right the model withholds, or stop the Engineer accepting.
    """
    required = REVIEW_PERMISSION[payload.decision]
    if not principal.has(required):
        raise PermissionDenied()

    try:
        return review_batch(
            session,
            batch_id=batch_id,
            organization_id=principal.organization_id,
            actor_id=principal.user_id,
            decision=payload.decision,
            note=payload.note,
        )
    except BatchNotFoundError as exc:
        raise _missing(exc) from exc
    except BatchStateError as exc:
        raise _conflict(exc) from exc
    except LaboratoryError as exc:
        # Segregation of duties: the executor may not review their own
        # batch. 403, because it is an authorization refusal about WHO,
        # not a malformed request.
        raise PermissionDenied(str(exc)) from exc
